### PR TITLE
Split out arrow-ipc

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #2594

# Rationale for this change
 

This will allow arrow-flight, parquet and potentially more to no longer need to depend on the full arrow crate

# What changes are included in this PR?

Splits out the arrow-ipc crate

# Are there any user-facing changes?

No breaking changes, there is a slight change in that CompressionCodec now contains all variants even when no features are enabled.

### CODE REVIEW

File: arrow-array/src/record_batch.rs [hunk 1]
```diff
@@ -22,6 +22,24 @@ use crate::{new_empty_array, Array, ArrayRef, StructArray};
 use arrow_schema::{ArrowError, DataType, Field, Schema, SchemaRef};
 use std::sync::Arc;
 
+/// Trait for types that can read `RecordBatch`'s.
+pub trait RecordBatchReader: Iterator<Item = Result<RecordBatch, ArrowError>> {
```
Comment (tustvold) [inline arrow-array/src/record_batch.rs:26]: This is moved from the arrow crate to allow using it in arrow-ipc and a future arrow-csv

File: arrow-ipc/src/compression.rs [hunk 2]
```diff
@@ -115,50 +117,85 @@ impl CompressionCodec {
...
+        output: &mut Vec<u8>,
+    ) -> Result<usize, ArrowError> {
+        match self {
+            CompressionCodec::Lz4Frame => decompress_lz4(input, output),
+            CompressionCodec::Zstd => decompress_zstd(input, output),
+        }
     }
 }
 
+#[cfg(feature = "lz4")]
```
Comment (tustvold) [inline arrow-ipc/src/compression.rs:141]: We now have separate features for the different compression codecs

File: arrow-ipc/src/writer.rs [hunk 3]
```diff
@@ -1503,18 +1501,18 @@ mod tests {
             ],
         )
         .unwrap();
-        let file_name = format!("target/debug/testdata/nulls_{}.arrow_file", suffix);
+        let mut file = tempfile::tempfile().unwrap();
```
Comment (tustvold) [inline arrow-ipc/src/writer.rs:1504]: This was failing about files not existing, I suspect some issue with moving to a different crate. Fortunately switching to use tempfile is cleaner and avoids this issue

File: arrow/tests/ipc_integration.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
+use arrow_ipc::reader::StreamReader;
+use arrow_ipc::writer::StreamWriter;
+use std::fs::File;
+use std::io::Seek;
+
+#[test]
+fn read_union_017() {
```
Comment (tustvold) [inline arrow/tests/ipc_integration.rs:7]: This test needed to be moved as test_util is still part of the top-level crate, this makes more sense as an integration test anyway so I think this is fine

File: arrow-ipc/src/writer.rs [hunk 5]
```diff
@@ -1626,45 +1610,6 @@ mod tests {
         assert!(dict_tracker.written.contains_key(&2));
     }
 
-    #[test]
```
Comment (tustvold) [inline arrow-ipc/src/writer.rs:1629]: Moved to an integration test

File: arrow/src/ipc/compression/stub.rs [hunk 6]
```diff
@@ -1,63 +0,0 @@
...
-
-//! Stubs that implement the same interface as the ipc_compression
-//! codec module, but always errors.
-
-use crate::buffer::Buffer;
-use crate::error::{ArrowError, Result};
-use crate::ipc::CompressionType;
-
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
-pub enum CompressionCodec {}
```
Comment (tustvold) [inline arrow/src/ipc/compression/stub.rs:26]: We no longer use a stub CompressionCodec, and instead generate runtime errors if the necessary features aren't enabled. This is consistent with how we handle this in parquet and how we handle other optional features such as timezone support

**Review by viirya — APPROVED** [review]
This looks good to me. Looks like it moves codes around with only necessary changes e.g. module names.

**Pull request merged**